Classes/common/Attachments/CDTBlobData.m

```diff
@@ -0,0 +1,206 @@
+
+#pragma mark - CDTBlobReader methods
+- (NSData *)dataWithError:(NSError **)error
+{
+    NSData *data = nil;
+    NSError *thisError = nil;
+
+    if ([self isBlobOpen]) {
+        CDTLogDebug(CDTDATASTORE_LOG_CONTEXT,
+                    @"Blob at %@ is open. Close it before reading its content", self.path);
+
+        thisError = [CDTBlobData errorOperationNotPossibleIfBlobIsOpen];
+    } else {
+        data = [NSData dataWithContentsOfFile:self.path
+                                      options:NSDataReadingMappedIfSafe
+                                        error:&thisError];
+        if (!data) {
+            CDTLogDebug(CDTDATASTORE_LOG_CONTEXT,
+                        @"Data object could not be created with file %@: %@", self.path, thisError);
+        }
+    }
+
+    if (!data && error) {
+        *error = thisError;
+    }
+
+    return data;
+}
+
+- (NSInputStream *)inputStreamWithOutputLength:(UInt64 *)outputLength
+{
+    if ([self isBlobOpen]) {
+        CDTLogDebug(CDTDATASTORE_LOG_CONTEXT, @"Close blob in order to create an input stream");
+
+        return nil;
+    }
+
+    NSFileManager *defaultManager = [NSFileManager defaultManager];
+    if (![defaultManager fileExistsAtPath:self.path isDirectory:NO]) {
+        CDTLogDebug(CDTDATASTORE_LOG_CONTEXT, @"No file found in %@", self.path);
+        
+        return nil;
+    }
+    
+    NSInputStream *inputStream = [NSInputStream inputStreamWithFileAtPath:self.path];
+
+    if (inputStream && outputLength) {
+        NSError *error = nil;
+        NSDictionary *info = [defaultManager attributesOfItemAtPath:self.path error:&error];
+        if (info) {
+            *outputLength = [info fileSize];
+        } else {
+            CDTLogDebug(CDTDATASTORE_LOG_CONTEXT,
+                        @"Attributes for file %@ could not be obtained: %@", self.path, error);
+
+            inputStream = nil;
+        }
+    }
+
+    return inputStream;
+}
+
+#pragma mark - CDTBlobWriter methods
+- (BOOL)createBlobWithData:(NSData *)data error:(NSError **)error
+{
+    BOOL success = NO;
+    NSError *thisError = nil;
+
+    if ([self isBlobOpen]) {
+        CDTLogDebug(CDTDATASTORE_LOG_CONTEXT,
+                    @"Blob at %@ is open. Close it before saving the data", self.path);
+
+        thisError = [CDTBlobData errorOperationNotPossibleIfBlobIsOpen];
+    } else {
+        NSDataWritingOptions options = NSDataWritingAtomic;
+#if TARGET_OS_IPHONE
+        options |= NSDataWritingFileProtectionCompleteUnlessOpen;
+#endif
+
+        success = [data writeToFile:self.path options:options error:&thisError];
+        if (!success) {
+            CDTLogDebug(CDTDATASTORE_LOG_CONTEXT, @"Could not write data to file %@: %@", self.path,
+                        thisError);
+        }
+    }
+
+    if (!success && error) {
+        *error = thisError;
+    }
+
+    return success;
+}
+
+- (BOOL)isBlobOpen { return (self.outFileHandle != nil); }
+
+- (BOOL)openBlobToAddData
```

Comment:
Should `Add` be `Append`? It's more specific, I'm a little unclear whether this method is for:

- Creating a new file;
- Appending an existing file;
- Overwriting an existing file.

Reading the protocol doc and the implementation has left me thinking perhaps this should be `Append`.